### DISCUSSION TITLE
Point cloud data to CF-NetCDF

### DISCUSSION THREAD

**Discussion by lhmarsden** [cf-convention/discuss#307] [Q&A about using CF]
### Question

I am trying to write a program to convert point cloud data to CF-NetCDF.

## What is a point cloud?

Imagine you have a structure - this could be a man-made structure like a building or a natural structure like a valley. A point cloud is a set of data points that repsent the 3D shape of that structure.

## Point clouds in CF-NetCDF

There are different ways to represent the coordinate system of a point cloud. For natural structures a global coordinate system is often used (latitude, longitude, elevation). I envisage that this type of data would be quite easy to store in a CF-NetCDF file, using one dimenion (a counter for each point) and variables for latitude, longitude, elevation.

But what about coordinate systems relative to some known coordinate - e.g. X, Y Z coordinates relative to the location of the scanner or some other user-defined location?

Of course we could encourage people to use global coordinate systems. However, sometimes the data are very high resolution (e.g. 1 mm) and scanned objects can be very small. I suspect it will be more efficient to use an X, Y, Z coordinate system relative to some reference point. 

**Question:** What is the best practice for using X, Y, Z coordinate systems relative to a defined point in CF-NetCDF? Consider that the files could be large, and storing latitude and longitude for every point might not be desirable.

Ideally the direction of X or Y should be defined too.

**Comment by lhmarsden** [comment]
Here is what I think would be suitable when using a global coordinate system:

```
netcdf test {
dimensions:
	point = 20 ;
variables:
	float longitude(point) ;
		longitude:units = "degrees_east" ;
		longitude:long_name = "longitude" ;
		longitude:standard_name = "longitude" ;
	float latitude(point) ;
		latitude:units = "degrees_north" ;
		latitude:long_name = "latitude" ;
		latitude:standard_name = "latitude" ;
	float altitude(point) ;
		altitude:units = "meters" ;
		altitude:long_name = "geometric height above geoid" ;
		altitude:standard_name = "altitude" ;
		altitude:positive = "up" ;
```

**Comment by JonathanGregory** [comment]
Dear Luke @lhmarsden 

Your example is certainly fine in CF. Your `point` dimension is an example of a discrete axis (sect 4.5) and could also be labelled as a `point` feature type (sect 9.1).

Best wishes

Jonathan

**Reply by lhmarsden** [reply]
Interesting, thanks @JonathanGregory, I had wrongly assumed that a `point` feature type could only include 1 point. I will use this!

**Comment by JonathanGregory** [comment]
Yes, it's fine to use the `point` feature type for a collection of points. However, apart from labelling it with `featureType`, there isn't any difference from what you did in your example, which is already fine in CF e.g. compare with example H.1. It makes perfect sense to have a discrete axis with several auxiliary coordinate variables. The discrete sampling geometries of chapter 9 are most useful with all the other types, which contain more than one element per feature, because of the extra machinery they provide for storing a collection of them with different sizes in a single data variable. For `point` features, one size (unity) fits all!

**Comment by JonathanGregory** [comment]
I've marked this with the `frequently asked question` label, although I don't remember it being asked before, because it might be helpful to copy it to the FAQ.
